packages/optimizer/lib/transformers/ApplyCommonAttributes.js

```diff
@@ -264,8 +260,8 @@ class ApplyCommonAttributes {
       return node.attribs.id;
     }
     node.attribs = node.attribs || [];
-    const id = ID_PREFIX + this.counter;
-    this.counter++;
+    const id = ID_PREFIX + this.transformedNodesCounter;
```

Comment:
The "next ID index" and the "number of nodes transformed by this transformer seems completely incidental to me and will only work correctly iff:
- no other transformer uses IDs with the same prefix;
- no preexisting IDs with the same prefix are found in the source document.

A more robust solution would be to have a separate object responsible for handing out IDs (that can be reused across transformers) and to check the next available index before using it.

I've solved this in the following way in the PHP version: https://github.com/ampproject/amp-toolbox-php/blob/v0.6.0/src/Dom/Document.php#L1806-L1849

Reply:
Note that in the PHP version, this is not a separate object (yet), but it is located in the `Dom\Document` abstraction, instead of within a transformer.

Reply:
I'm checking for existing ids here: https://github.com/ampproject/amp-toolbox/blob/main/packages/optimizer/lib/transformers/ApplyCommonAttributes.js#L265. The counter will increase until there's a free id. 

Moving this into a separate service is definitely the best approach, but I'd only add this when we actually need it.

Reply:
Ah, yes, I see. You're collecting the IDs in a `Set` to do lookups. Indeed, this will need to be extracted out when it is needed in another place, but only when that is actually the case.

This made me look into the performance tradeoffs of my approach vs yours in terms of checking existing IDs, though. I've been using `getElementById()`, assuming that would be a sort of hashed lookup. But reading up on it, I'm not so sure anymore, as this is meant to work with multiple, freely definable and namespaceable ID attributes in XML.

I'll have to test whether it might be more efficient to keep an up-to-date lookup table like yours within the `Dom\Document` abstraction...

Reply:
I went with the Set as we don't have the luxury of a `getElementById` API and we're already traversing over all nodes...